Support for robots using PCA 9685 motor controller

neworderofjamie opened the pull request:
This is basically a chip which sits on the I2C bus and controls 16 Pulse Width Modulation channels which can be used be used to drive motors/flashing LEDS/whatever. 

You can get boards containing the chip by itself (https://thepihut.com/products/adafruit-16-channel-12-bit-pwm-servo-driver-i2c-interface-pca9685) or with motor drivers attached (https://thepihut.com/products/adafruit-dc-motor-stepper-featherwing-add-on-for-all-feather-boards). In this PR we:
* Define a simple interface to the PCA9685 in common 
* Define a wrapper around this to control the featherwing 
* However, turns out, the featherwing motor drivers are particularly weedy and the mecanum motors burns them out. On the new mecanum we've instead hooked up 4 motor driver boards to the first 8 channels of a PCA9685 and this configuration is supported by the ``MecanumPCA9685`` robot type.

alexdewar commented on include/common/pca_9685.h:
```diff
@@ -0,0 +1,104 @@
+        CHANNEL_13 = 13,
+        CHANNEL_14 = 14,
+        CHANNEL_15 = 15,
+    };
```
I feel like having this as an ``enum`` is a bit unnecessary, but it's harmless I guess

neworderofjamie replied:
It just means you don't need to assert somewhere that channel >= 1 and channel <= 15 :man_shrugging:

alexdewar commented on src/common/pca_9685.cc:
```diff
@@ -0,0 +1,138 @@
+void PCA9685::setDutyCycle(Channel channel, float dutyCycle)
+{
+    // Clamp duty cycle
+    dutyCycle = std::min(1.0f, std::max(0.0f, dutyCycle));
```
Elsewhere in BoB robotics I've generally used error checks rather than silently capping values. The danger is that your code ends up passing some nonsense value in and you never notice.

How about:
```c++
BOB_ASSERT(dutyCycle >= 0.f && dutyCycle <= 1.f);
```

alexdewar commented on src/robots/mecanum_pca_9685.cc:
```diff
@@ -0,0 +1,85 @@
+    cap(m1);
+    cap(m2);
+    cap(m3);
+    cap(m4);
```
See comment above re capping (although I have a feeling we may actually do this in the old mecanum code)

neworderofjamie replied:
I wonder here if there should be an assert here and capping in the omni2D instead as I fear the derived m1, m2, m3 and m4 values may well end up not being between -1 and 1

alexdewar commented on src/robots/mecanum_pca_9685.cc:
```diff
@@ -0,0 +1,85 @@
+    setMotorThrottle(Motor::MOTOR_4, m4);
+}
+//----------------------------------------------------------------------------
+void MecanumPCA9685::setMotorThrottle(Motor motor, float throttle)
```
The problem with ``Motor`` being an ``enum`` rather than an ``enum class`` is that a user could accidentally pass e.g. a value of ``5`` for ``motor`` and the compiler would happily accept it. (You could also just check for this case with ``BOB_ASSERT()``).

neworderofjamie replied:
True - as it's an old-style enum I should check - bleurgh

alexdewar reviewed (commented):
LGTM aside from a couple of ``enum``-related niggles!

neworderofjamie commented on include/common/dc_motor_featherwing.h:
```diff
@@ -0,0 +1,66 @@
+        MOTOR_CHANNEL_POSITIVE,
+        MOTOR_CHANNEL_NEGATIVE,
+        MOTOR_CHANNEL_MAX
+    };
```
Because I'm using them (in a somewhat old school way) to index into ``motorChannels`` which enum classes make super irritating (``static_cast`` EVERYWHERE)

neworderofjamie commented:
I've added asserts to cover the various enum and floating point input limits and moved the capping in the mecanum. Assuming I haven't broken it on the robot, are you ok to merge?

alexdewar commented:
Yep!

I'd also like to get #257 merged as I'm using it as a base for something else I'm doing. Are you ok to merge that once I've devirtualised `MecanumPCA9685` in my branch?

neworderofjamie commented:
So you've already merged this into #257?

neworderofjamie commented:
Ok, joystick_robot doesn't hit any asserts so I'm gonna merge
